pkg/cmd/infra/router/template.go

```diff
@@ -2,87 +2,162 @@ package router
-	flag.StringVar(&cfg.TemplateRouterConfig.TemplateFile, "template", util.Env("TEMPLATE_FILE", ""), "The path to the template file to use")
-	flag.StringVar(&cfg.TemplateRouterConfig.ReloadScript, "reload", util.Env("RELOAD_SCRIPT", ""), "The path to the reload script to use")
-	flag.StringVar(&cfg.TemplateRouterConfig.StatsPort, "stats-port", util.Env("STATS_PORT", ""), "If the underlying router implementation can provide statistics this is a hint to expose it on this port.")
-	flag.StringVar(&cfg.TemplateRouterConfig.StatsPassword, "stats-password", util.Env("STATS_PASSWORD", ""), "If the underlying router implementation can provide statistics this is the requested password for auth.")
-	flag.StringVar(&cfg.TemplateRouterConfig.StatsUsername, "stats-user", util.Env("STATS_USERNAME", ""), "If the underlying router implementation can provide statistics this is the requested username for auth.")
+func (o *RouterStats) Bind(flag *pflag.FlagSet) {
+	flag.StringVar(&o.StatsPortString, "stats-port", util.Env("STATS_PORT", ""), "If the underlying router implementation can provide statistics this is a hint to expose it on this port.")
+	flag.StringVar(&o.StatsPassword, "stats-password", util.Env("STATS_PASSWORD", ""), "If the underlying router implementation can provide statistics this is the requested password for auth.")
+	flag.StringVar(&o.StatsUsername, "stats-user", util.Env("STATS_USERNAME", ""), "If the underlying router implementation can provide statistics this is the requested username for auth.")
 }
 
-// makeTemplatePlugin creates a template router plugin.
-func makeTemplatePlugin(cfg *routerConfig) (*templateplugin.TemplatePlugin, error) {
-	if cfg.TemplateRouterConfig.TemplateFile == "" {
-		return nil, errors.New("Template file must be specified")
+// NewCommndTemplateRouter provides CLI handler for the template router backend
+func NewCommandTemplateRouter(name string) *cobra.Command {
+	options := &TemplateRouterOptions{
+		Config: clientcmd.NewConfig(),
+	}
+	options.Config.FromFile = true
+
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s%s", name, clientcmd.ConfigSyntax),
+		Short: "Start a router",
+		Long:  routerLong,
+		Run: func(c *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.Complete())
+			cmdutil.CheckErr(options.Validate())
+			cmdutil.CheckErr(options.Run())
+		},
 	}
 
-	if cfg.TemplateRouterConfig.ReloadScript == "" {
-		return nil, errors.New("Reload script must be specified")
+	cmd.AddCommand(version.NewVersionCommand(name))
+
+	flag := cmd.Flags()
+	options.Config.Bind(flag)
+	options.TemplateRouter.Bind(flag)
+	options.RouterSelection.Bind(flag)
+
+	return cmd
+}
+
+func (o *TemplateRouterOptions) Complete() error {
+	routerSvcName := util.Env("ROUTER_SERVICE_NAME", "")
+	routerSvcNamespace := util.Env("ROUTER_SERVICE_NAMESPACE", "")
+	if len(routerSvcName) > 0 {
+		if len(routerSvcNamespace) == 0 {
+			return fmt.Errorf("ROUTER_SERVICE_NAMESPACE is required when ROUTER_SERVICE_NAME is specified")
```

Comment:
Things we pull from Env come from Complete

On Mon, Aug 31, 2015 at 7:07 PM, Ram Ranganathan <notifications@github.com>
wrote:

> In pkg/cmd/infra/router/template.go
> <https://github.com/openshift/origin/pull/4428#discussion_r38365620>:
>
> > +	cmd.AddCommand(version.NewVersionCommand(name))
> > +
> > +	flag := cmd.Flags()
> > +	options.Config.Bind(flag)
> > +	options.TemplateRouter.Bind(flag)
> > +	options.RouterSelection.Bind(flag)
> > +
> > +	return cmd
> > +}
> > +
> > +func (o *TemplateRouterOptions) Complete() error {
> > +	routerSvcName := util.Env("ROUTER_SERVICE_NAME", "")
> > +	routerSvcNamespace := util.Env("ROUTER_SERVICE_NAMESPACE", "")
> > +	if len(routerSvcName) > 0 {
> > +		if len(routerSvcNamespace) == 0 {
> > +			return fmt.Errorf("ROUTER_SERVICE_NAMESPACE is required when ROUTER_SERVICE_NAME is specified")
>
> Any reason this code is in Complete vis-a-vis in Validate below (as it is
> a validation)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4428/files#r38365620>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift